### PR TITLE
Remove guice from Mapper plugins

### PR DESCRIPTION
**Pull request by rjernst**
This changes adds a MapperPlugin interface which allows pull style
retrieval of mappers and metadata mappers added by plugins. For now, I
have kept the MapperRegistry, but this should be removed in the future
as it is just a silly container for 2 maps which could themselves be
passed around.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/indices/IndicesModule.java [hunk 1]
```diff
@@ -92,65 +94,63 @@ private void registerBuildInWritables() {
...
+        mappers.put(ObjectMapper.CONTENT_TYPE, new ObjectMapper.TypeParser());
+        mappers.put(ObjectMapper.NESTED_CONTENT_TYPE, new ObjectMapper.TypeParser());
+        mappers.put(CompletionFieldMapper.CONTENT_TYPE, new CompletionFieldMapper.TypeParser());
+        mappers.put(GeoPointFieldMapper.CONTENT_TYPE, new GeoPointFieldMapper.TypeParser());
         if (ShapesAvailability.JTS_AVAILABLE && ShapesAvailability.SPATIAL4J_AVAILABLE) {
-            registerMapper(GeoShapeFieldMapper.CONTENT_TYPE, new GeoShapeFieldMapper.TypeParser());
+            mappers.put(GeoShapeFieldMapper.CONTENT_TYPE, new GeoShapeFieldMapper.TypeParser());
         }
+
+        for (MapperPlugin mapperPlugin : mapperPlugins) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/indices/IndicesModule.java]: should we make sure that they have an empty intersection? `Sets.hasEmptyIntersection(mappers.keySet(), mapperPlugin.getMappers().keySet()`?

File: core/src/main/java/org/elasticsearch/indices/IndicesModule.java [hunk 2]
```diff
@@ -92,65 +94,63 @@ private void registerBuildInWritables() {
...
+        metadataMappers.put(SourceFieldMapper.NAME, new SourceFieldMapper.TypeParser());
+        metadataMappers.put(TypeFieldMapper.NAME, new TypeFieldMapper.TypeParser());
+        metadataMappers.put(AllFieldMapper.NAME, new AllFieldMapper.TypeParser());
+        metadataMappers.put(TimestampFieldMapper.NAME, new TimestampFieldMapper.TypeParser());
+        metadataMappers.put(TTLFieldMapper.NAME, new TTLFieldMapper.TypeParser());
+        metadataMappers.put(VersionFieldMapper.NAME, new VersionFieldMapper.TypeParser());
+        metadataMappers.put(ParentFieldMapper.NAME, new ParentFieldMapper.TypeParser());
+        // _field_names is not registered here, see below
+
+        for (MapperPlugin mapperPlugin : mapperPlugins) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/indices/IndicesModule.java]: same here should we cehck for empty intersection?

**Comment by s1monw** [comment]
left two comments - LGTM otherwise

**Comment by jpountz** [comment]
LGTM

**Comment by clintongormley** [comment]
@rjernst should this be "enhancement" or "breaking"?

**Comment by rjernst** [comment]
@clintongormley I added the `breaking` label, and added a note to the migration guide.

**Pull request merged**